### PR TITLE
Add actions to response when the user denied to turn off hwdec

### PR DESCRIPTION
**Pull request by uiryuu**
- [x] This change has been discussed with the author.
- [ ] It implements / fixes issue #.

---

**Description:**

### CODE REVIEW

File: iina/MainMenuActions.swift [hunk 1]
```diff
@@ -180,7 +180,7 @@ extension MainMenuActionHandler {
 
   @objc func menuChangeCrop(_ sender: NSMenuItem) {
     if let cropStr = sender.representedObject as? String {
-      player.setCrop(fromString: cropStr)
+      let _ = player.setCrop(fromString: cropStr)
```
Comment (saagarjha) [inline iina/MainMenuActions.swift]: I believe you don't need the `let` here if you're just trying to discard the result.

File: iina/PlayerCore.swift [hunk 2]
```diff
@@ -667,30 +667,40 @@ class PlayerCore: NSObject {
...
       if let filter = info.cropFilter {
         let _ = removeVideoFilter(filter)
         info.unsureCrop = "None"
       }
     }
+    return nil
   }
 
-  func setCrop(fromFilter filter: MPVFilter) {
+  func setCrop(fromFilter filter: MPVFilter) -> Bool {
```
Comment (saagarjha) [inline iina/PlayerCore.swift]: You may also want to consider making this `@discardableResult` if you plan on ignoring the return value often.

File: iina/PlayerCore.swift [hunk 3]
```diff
@@ -706,7 +716,12 @@ class PlayerCore: NSObject {
     }
   }
 
-  func addVideoFilter(_ filter: MPVFilter) -> Bool {
+  /**
+   Try to add a video filter.
+
+   - Returns: `true` for success; `false` for user abort; `nil` for a failure from mpv.
+   */
+  func addVideoFilter(_ filter: MPVFilter) -> Bool? {
```
Comment (saagarjha) [inline iina/PlayerCore.swift]: Similarly you may consider making this `throw` as well

File: iina/DraggingDetect.swift [hunk 4]
```diff
@@ -221,10 +221,10 @@ extension PlayerCore {
           "file": paths[0],
```
Comment (saagarjha) [inline iina/DraggingDetect.swift]: `let result = addVideoFilter(…) ?? false`

File: iina/FilterWindowController.swift [hunk 5]
```diff
@@ -150,7 +146,7 @@ class FilterWindowController: NSWindowController {
     let filter = savedFilters[row]
     let pc = PlayerCore.active
     if sender.state == .on {
-      if pc.addVideoFilter(MPVFilter(rawString: filter.filterString)!) {
+      if let success = pc.addVideoFilter(MPVFilter(rawString: filter.filterString)!), success {
```
Comment (saagarjha) [inline iina/FilterWindowController.swift]: You're not using `success`, so why not just:

```swift
if pc.addVideoFilter(MPVFilter(rawString: filter.filterString)!) ?? false {
	…
}
```

File: iina/FilterWindowController.swift [hunk 6]
```diff
@@ -403,8 +399,15 @@ class NewFilterSheetViewController: NSViewController, NSTableViewDelegate, NSTab
       }
     }
     // create filter
-    filterWindow.addFilter(preset.transformer(instance))
-    PlayerCore.active.sendOSD(.addFilter(preset.localizedName))
+    let filter = preset.transformer(instance)
+    if let success = PlayerCore.active.addVideoFilter(filter) {
```
Comment (saagarjha) [inline iina/FilterWindowController.swift]: Ditto

File: iina/FreeSelectingViewController.swift [hunk 7]
```diff
@@ -24,11 +24,13 @@ class FreeSelectingViewController: CropBoxViewController {
       if let existingFilter = player.info.delogoFiter {
         let _ = player.removeVideoFilter(existingFilter)
       }
-      if !player.addVideoFilter(filter) {
-        Utility.showAlert("filter.incorrect")
-        return
+      if let success = player.addVideoFilter(filter) {
```
Comment (saagarjha) [inline iina/FreeSelectingViewController.swift]: Same as above

File: iina/MainWindowMenuActions.swift [hunk 8]
```diff
@@ -177,7 +177,7 @@ extension MainWindowController {
       } else {
         // add
         if isVideo {
-          if !player.addVideoFilter(filter) {
+          if let success = player.addVideoFilter(filter), !success {
```
Comment (saagarjha) [inline iina/MainWindowMenuActions.swift]: Again, you're not using `success` anywhere so you can discard it

File: iina/PlayerCore.swift [hunk 9]
```diff
@@ -509,7 +509,7 @@ class PlayerCore: NSObject {
       if info.flipFilter == nil {
         let vf = MPVFilter.flip()
         vf.label = Constants.FilterName.flip
-        if addVideoFilter(vf) {
+        if let success = addVideoFilter(vf), success {
```
Comment (saagarjha) [inline iina/PlayerCore.swift]: Same

File: iina/PlayerCore.swift [hunk 10]
```diff
@@ -526,7 +526,7 @@ class PlayerCore: NSObject {
       if info.mirrorFilter == nil {
         let vf = MPVFilter.mirror()
         vf.label = Constants.FilterName.mirror
-        if addVideoFilter(vf) {
+        if let success = addVideoFilter(vf), success {
```
Comment (saagarjha) [inline iina/PlayerCore.swift]: "

File: iina/PlayerCore.swift [hunk 11]
```diff
@@ -667,30 +667,40 @@ class PlayerCore: NSObject {
     syncUITime()
   }
 
-  func setCrop(fromString str: String) {
+  func setCrop(fromString str: String) -> Bool? {
```
Comment (saagarjha) [inline iina/PlayerCore.swift]: I don't see you trying to handle `nil` anywhere, so why not make the function `throw` an error instead?

File: iina/PlayerCore.swift [hunk 12]
```diff
@@ -667,30 +667,40 @@ class PlayerCore: NSObject {
...
     let vwidth = info.videoWidth!
     let vheight = info.videoHeight!
     if let aspect = Aspect(string: str) {
       let cropped = NSMakeSize(CGFloat(vwidth), CGFloat(vheight)).crop(withAspect: aspect)
       let vf = MPVFilter.crop(w: Int(cropped.width), h: Int(cropped.height), x: nil, y: nil)
       vf.label = Constants.FilterName.crop
-      setCrop(fromFilter: vf)
+      if !setCrop(fromFilter: vf) {
```
Comment (saagarjha) [inline iina/PlayerCore.swift]: Stylistically I think a `guard` better conveys your intent

File: iina/PlayerCore.swift [hunk 13]
```diff
@@ -667,30 +667,40 @@ class PlayerCore: NSObject {
...
     }
+    return nil
   }
 
-  func setCrop(fromFilter filter: MPVFilter) {
+  func setCrop(fromFilter filter: MPVFilter) -> Bool {
     filter.label = Constants.FilterName.crop
-    if addVideoFilter(filter) {
-      info.cropFilter = filter
+    if let success = addVideoFilter(filter) {
```
Comment (saagarjha) [inline iina/PlayerCore.swift]: `success` isn't used again

File: iina/QuickSettingViewController.swift [hunk 14]
```diff
@@ -401,8 +401,11 @@ class QuickSettingViewController: NSViewController, NSTableViewDataSource, NSTab
 
   @IBAction func cropChangedAction(_ sender: NSSegmentedControl) {
     let cropStr = AppData.cropsInPanel[sender.selectedSegment]
-    player.setCrop(fromString: cropStr)
-    mainWindow.displayOSD(.crop(cropStr))
+    if let success = player.setCrop(fromString: cropStr), !success {
```
Comment (saagarjha) [inline iina/QuickSettingViewController.swift]: `success` isn't necessary

**Review by saagarjha — COMMENTED** [review]
I know you didn't ask for my review, but here's a couple things I saw when glancing over. It's your choice whether you want to implement them.

**Comment by alejx** [comment]
@saagarjha Thanks for your review. I would like to revise to make the function `throws`.

File: iina/MainMenuActions.swift (hunk 1 above)
Comment (alejx) [reply iina/MainMenuActions.swift]: We should fix it in a future commit since `let _ = ...` also exists in other parts of our codebase.

**Comment by lhc70000** [comment]
There're still a lot of problems in this PR. Closing for now.

**Pull request closed without merging**